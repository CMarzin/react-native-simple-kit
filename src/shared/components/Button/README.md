# Button ðŸ”˜

A generic button component that handles simple functions or navigation paths into React Router v4 based apps.

### API

* color?: `'blue' |Â 'green' | 'red' | 'yellow' | 'white'`
* children?: `React$Element<any>`
* inactive?: `boolean`
* onLongPress?: `Function`
* onPress: `Function`
* style?: `StyleSheet`
* text?: `string`

### Example

```js
import React from 'react'
import { SafeAreaView } from 'react-native'
import Button from '@components/Button'

const MyComponent = (): React$Element<any> => (
  <SafeAreaView style={{ flex: 1 }}>
    <Button
      color="red"
      onPress={() => props.navigator.popToRoot({
        animated: true,
        animationType: 'fade'
      })}
      text="Go back"
    />
  </SafeAreaView>
)

export default MyComponent
```

## Changelog
#### Tuesday, January 23rd, 2018
- Matching API with new routing lib.
#### Wednesday, November 22th, 2017
- Switch to `React.PureComponent`.

#### Monday, November 20th, 2017
- Add `blue` as a supported color.

#### Thursday, November 16th, 2017
- Initial release. We didn't track changes before this version.
