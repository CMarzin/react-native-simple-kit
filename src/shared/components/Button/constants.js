/* @flow */

import {
  BRAND_COLOR_BLUE,
  BRAND_COLOR_GREEN,
  BRAND_COLOR_RED,
  BRAND_COLOR_YELLOW,
  NEUTRAL_COLOR_00,
} from '@theme/colors'

export type DefaultProps = {
  color?: 'blue' | 'green' | 'red' | 'yellow' | 'white',
  children?: ?React$Element<*>,
  inactive?: ?boolean,
  onLongPress?: Function,
  style?: ?StyleSheet,
  text?: ?string,
}

export type Props = DefaultProps & {
  onPress: Function,
}

export const brandColors = {
  blue: BRAND_COLOR_BLUE,
  green: BRAND_COLOR_GREEN,
  red: BRAND_COLOR_RED,
  yellow: BRAND_COLOR_YELLOW,
  white: NEUTRAL_COLOR_00,
}
