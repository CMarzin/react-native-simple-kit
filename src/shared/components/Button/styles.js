/* @flow */

import { StyleSheet } from 'react-native'
import { rem } from '@helpers/responsive'
import { SUCCESS_COLOR, DEFAULT_TEXT_COLOR } from '@theme/colors'
import { AVENIR_MEDIUM } from '@theme/fonts'

const styles: Object = StyleSheet.create({
  container: {
    minWidth: rem(160),
    minHeight: rem(42),
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: SUCCESS_COLOR,
    backgroundColor: SUCCESS_COLOR,
    borderRadius: 100,
  },
  text: {
    ...AVENIR_MEDIUM,
    fontSize: rem(16),
    textAlign: 'center',
    color: DEFAULT_TEXT_COLOR,
    backgroundColor: 'transparent',
  },
})

export default styles
