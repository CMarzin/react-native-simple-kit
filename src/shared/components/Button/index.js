/* @flow */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Text, View } from 'react-native'
import { Link } from 'react-router-native'

import { DEFAULT_TEXT_COLOR_DARK } from '@theme/colors'
import Touchable from '@components/Touchable'

import {
  type ButtonProps as Props,
  type DefaultProps,
  brandColors,
} from './constants'
import styles from './styles'

class Button extends Component<Props, void> {
  props: Props

  static propTypes = {
    children: PropTypes.element,
    color: PropTypes.oneOf(['green', 'red', 'yellow', 'white']),
    inactive: PropTypes.bool,
    onPress: PropTypes.func,
    to: PropTypes.string,
    replace: PropTypes.bool,
    routeState: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    text: PropTypes.string,
  }

  static defaultProps: DefaultProps = {
    children: undefined,
    color: 'green',
    inactive: false,
    onPress: undefined,
    to: undefined,
    replace: false,
    routeState: undefined,
    style: undefined,
    text: undefined,
  }

  componentWillMount() {
    if (!this.props.children && !this.props.text) {
      throw new Error('No children or `text` prop was passed to <Button />')
    }
    if (!this.props.to && !this.props.onPress) {
      throw new Error('No `to` path or `onPress()` function was passed to <Button />')
    }
  }

  render() {
    const {
      color,
      children,
      inactive,
      onPress,
      replace,
      routeState,
      style,
      to,
      text,
    } = this.props
    return (
      <Link
        onPress={onPress}
        to={{ pathname: to || '', state: routeState || {} }}
        component={Touchable}
        replace={replace || false}
      >
        <View
          style={[
            styles.container,
            color && {
              backgroundColor: brandColors[color],
              borderColor: brandColors[color],
            },
            inactive && {
              backgroundColor: color && 'transparent',
              borderColor: color && brandColors[color],
            },
            style,
          ]}
        >
          {text &&
            <Text
              style={[
                styles.text,
                color === 'white' && { color: DEFAULT_TEXT_COLOR_DARK },
                inactive && { color: color && brandColors[color] },
              ]}
            >
              {text}
            </Text>}
          {children}
        </View>
      </Link>
    )
  }
}

export default Button
