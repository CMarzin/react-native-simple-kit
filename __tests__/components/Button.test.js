import React from 'react'
import TestRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { Text } from 'react-native'
import { NativeRouter } from 'react-router-native'
import Button from '@components/Button'

describe('Generic component: <Button />', () => {
  it('renders correctly', () => {
    const component = TestRenderer.create(
      <NativeRouter>
        <Button
          onPress={() => console.log('Success ✅')}
          text="Hello!"
        />
      </NativeRouter>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls onPress when pressed', () => {
    const renderer = new ShallowRenderer()
    const onPress = jest.fn()

    renderer.render((
      <Button
        onPress={() => onPress()}
        text="Hello!"
      />
    ))

    const output = renderer.getRenderOutput()
    output.props.onPress()

    expect(onPress).toBeCalled()
  })

  it('renders children if passed', () => {
    const component = TestRenderer.create(
      <NativeRouter>
        <Button
          onPress={() => console.log('Success ✅')}
        >
          <Text>Hello!</Text>
        </Button>
      </NativeRouter>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
