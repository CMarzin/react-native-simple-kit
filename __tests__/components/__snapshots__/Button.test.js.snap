// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic component: <Button /> renders children if passed 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <RCTSafeAreaView
    style={null}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#53C000",
            "borderColor": "#53C000",
            "borderRadius": 100,
            "borderWidth": 1,
            "justifyContent": "center",
            "minHeight": 43.58681250000001,
            "minWidth": 166.04500000000002,
          },
          Object {
            "backgroundColor": "#53C000",
            "borderColor": "#53C000",
          },
          false,
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        Hello!
      </Text>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`Generic component: <Button /> renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <RCTSafeAreaView
    style={null}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#53C000",
            "borderColor": "#53C000",
            "borderRadius": 100,
            "borderWidth": 1,
            "justifyContent": "center",
            "minHeight": 43.58681250000001,
            "minWidth": 166.04500000000002,
          },
          Object {
            "backgroundColor": "#53C000",
            "borderColor": "#53C000",
          },
          false,
          undefined,
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#fefefe",
              "fontFamily": "Avenir",
              "fontSize": 16.6045,
              "fontWeight": "600",
              "textAlign": "center",
            },
            false,
            false,
          ]
        }
      >
        Hello!
      </Text>
    </View>
  </RCTSafeAreaView>
</View>
`;
